File: ACO-template-Java/ci-assignment3-template-master 2/src/Maze.java
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

/**
 * Class that holds all the maze data. This means the pheromones, the open and blocked tiles in the system as
 * well as the starting and end coordinates.
 */
public class Maze {
    public int width;
    public int length;
    public double[][] walls;
    public double[][] pheromones;
    public Coordinate start;
    public Coordinate end;

    /**
     * Constructor of a maze
     * @param walls int array of tiles accessible (1) and non-accessible (0)
     * @param width width of Maze (horizontal)
     * @param length length of Maze (vertical)
     */
    public Maze(int[][] wallsw, int width, int length) {


        this.length = length;
        this.width = width;
        this.walls = new double[width][length];
        this.pheromones = new double[width][length];
        for (int row = 0; row < width; row++) {
            for (int column = 0; column < length; column++) {
                this.walls[row][column] = (double) wallsw[row][column];
            }
        }
        initializePheromones();
    }

    /**
     * Initialize pheromones to a start value.
     */
    private void initializePheromones() {
        for (int row = 0; row < pheromones.length; row++) {
            for (int column = 0; column < pheromones[0].length; column++) {
                this.pheromones[row][column] = this.walls[row][column];
            }
        }
    }

    /**
     * Reset the maze for a new shortest path problem.
     */
    public void reset() {
        initializePheromones();
    }

    /**
     * Update the pheromones along a certain route according to a certain Q
     * @param r The route of the ants
     * @param Q Normalization factor for amount of dropped pheromone
     */
    public void addPheromoneRoute(Route r, double Q) throws Exception {
        Coordinate curPos = r.start;
        int routeSize = r.size();
        for (Direction dir : r.route) {
            curPos.add(dir);
            if (routeSize == 0)
                throw new Exception("sad");
            this.pheromones[curPos.getX()][curPos.getY()] += Q * (1.0 / routeSize);

        }
    }

    /**
     * Update pheromones for a list of routes
     * @param routes A list of routes
     * @param Q Normalization factor for amount of dropped pheromone
     */
    public void addPheromoneRoutes(List<Route> routes, double Q) throws Exception {
        for (Route r : routes) {
            addPheromoneRoute(r, Q);
        }
    }

    /**
     * Evaporate pheromone
     * @param rho evaporation factor
     */
    public void evaporate(double rho) {
        for (int row = 0; row < pheromones.length; row++) {
            for (int column = 0; column < pheromones[0].length; column++) {
                this.pheromones[row][column] = this.pheromones[row][column] * (1-rho);
            }
        }
    }

    /**
     * Width getter
     * @return width of the maze
     */
    public int getWidth() {
        return width;
    }

    /**
     * Length getter
     * @return length of the maze
     */
    public int getLength() {
        return length;
    }


    /**
     * Returns a the amount of pheromones on the neighbouring positions (N/S/E/W).
     * @param position The position to check the neighbours of.
     * @return the pheromones of the neighbouring positions.
     */
    public SurroundingPheromone getSurroundingPheromone(Coordinate position) {
        double east = 0,west= 0,north= 0,south = 0;

        Coordinate dir = position.add(Direction.East);
        if(inBounds(dir)){
            east = pheromones[dir.getX()][dir.getY()];
        }

        dir = position.add(Direction.West);
        if(inBounds(dir)){
            west = pheromones[dir.getX()][dir.getY()];
        }
        dir = position.add(Direction.North);
        if(inBounds(dir)){
            north = pheromones[dir.getX()][dir.getY()];
        }
        dir = position.add(Direction.South);
        if(inBounds(dir)){
            south = pheromones[dir.getX()][dir.getY()];
        }
        return new SurroundingPheromone(north,east,south,west);
    }


    /**
     * Check whether a coordinate lies in the current maze.
     * @param position The position to be checked
     * @return Whether the position is in the current maze
     */
    private boolean inBounds(Coordinate position) {
        return position.xBetween(0, width) && position.yBetween(0, length);
    }

    /**
     * Representation of Maze as defined by the input file format.
     * @return String representation
     */
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(width);
        sb.append(' ');
        sb.append(length);
        sb.append(" \n");
        for (int y = 0; y < length; y++) {
            for (int x = 0; x < width; x++ ) {
                sb.append(walls[x][y]);
                sb.append(' ');
            }
            sb.append("\n");
        }
        return sb.toString();
    }

    /**
     * Method that builds a mze from a file
     * @param filePath Path to the file
     * @return A maze object with pheromones initialized to 0's inaccessible and 1's accessible.
     */
    public static Maze createMaze(String filePath) throws FileNotFoundException {
        Scanner scan = new Scanner(new FileReader(filePath));
        int width = scan.nextInt();
        int length = scan.nextInt();
        int[][] mazeLayout = new int[width][length];
        for (int y = 0; y < length; y++) {
            for (int x = 0; x < width; x++) {
                mazeLayout[x][y] = scan.nextInt();
            }
        }
        scan.close();
        return new Maze(mazeLayout, width, length);
    }
}

File: ACO-Python-master/src/Maze.py
import os, sys

sys.path.append(os.path.dirname(os.path.dirname(os.path.realpath(__file__))))

import traceback
import SurroundingPheromone as sr
from Direction import Direction
import numpy as np


# Class that holds all the maze data. This means the pheromones, the open and blocked tiles in the system as
# well as the starting and end coordinates.
class Maze:

    # Constructor of a maze
    # @param walls int array of tiles accessible (1) and non-accessible (0)
    # @param width width of Maze (horizontal)
    # @param length length of Maze (vertical)
    def __init__(self, walls, width, length):
        self.pheromone_array = None
        self.walls = walls
        self.length = length
        self.width = width
        self.start = None
        self.end = None
        self.initialize_pheromones()

    # Initialize pheromones to a start value.
    def initialize_pheromones(self):
        self.pheromone_array = np.array(self.walls)
        return self.pheromone_array

    # Reset the maze for a new shortest path problem.
    def reset(self):
        self.initialize_pheromones()

    # Update the pheromones along a certain route according to a certain Q
    # @param r The route of the ants
    # @param Q Normalization factor for amount of dropped pheromone
    def add_pheromone_route(self, route, q):
        current_pos = route.get_start()
        for direction in route.get_route():
            current_pos = current_pos.add_direction(direction)
            x = current_pos.get_x()
            y = current_pos.get_y()
            self.pheromone_array[x][y] += q * (1/route.size())
        return self.pheromone_array

     # Update pheromones for a list of routes
     # @param routes A list of routes
     # @param Q Normalization factor for amount of dropped pheromone
    def add_pheromone_routes(self, routes, q):
        for r in routes:
            self.add_pheromone_route(r, q)

    # Evaporate pheromone
    # @param rho evaporation factor
    def evaporate(self, rho):
        return self.pheromone_array*rho

    # Width getter
    # @return width of the maze
    def get_width(self):
        return self.width

    # Length getter
    # @return length of the maze
    def get_length(self):
        return self.length

    # Returns the amount of pheromones on the neighbouring positions (N/S/E/W).
    # @param position The position to check the neighbours of.
    # @return the pheromones of the neighbouring positions.
    def get_surrounding_pheromone(self, position):
        north = 0
        west = 0
        south = 0
        east = 0
        direction = position.add_direction(Direction.north)
        if self.in_bounds(direction):
            north = self.pheromone_array[direction.get_x()][direction.get_y()]
        direction = position.add_direction(Direction.east)
        if self.in_bounds(direction):
            east = self.pheromone_array[direction.get_x()][direction.get_y()]
        direction = position.add_direction(Direction.south)
        if self.in_bounds(direction):
            south = self.pheromone_array[direction.get_x()][direction.get_y()]
        direction = position.add_direction(Direction.west)
        if self.in_bounds(direction):
            west = self.pheromone_array[direction.get_x()][direction.get_y()]
        return sr.SurroundingPheromone(north, east, south, west)

    # Pheromone getter for a specific position. If the position is not in bounds returns 0
    # @param pos Position coordinate
    # @return pheromone at point
    def get_pheromone(self, pos):
        return self.pheromone_array[pos.get_y, pos.get_x]

    # Check whether a coordinate lies in the current maze.
    # @param position The position to be checked
    # @return Whether the position is in the current maze
    def in_bounds(self, position):
        return position.x_between(0, self.width) and position.y_between(0, self.length)

    # Representation of Maze as defined by the input file format.
    # @return String representation
    def __str__(self):
        string = ""
        string += str(self.width)
        string += " "
        string += str(self.length)
        string += " \n"
        for y in range(self.length):
            for x in range(self.width):
                string += str(self.walls[x][y])
                string += " "
            string += "\n"
        return string

    # Method that builds a mze from a file
    # @param filePath Path to the file
    # @return A maze object with pheromones initialized to 0's inaccessible and 1's accessible.
    @staticmethod
    def create_maze(file_path):
        try:
            f = open(file_path, "r")
            lines = f.read().splitlines()
            dimensions = lines[0].split(" ")
            width = int(dimensions[0])
            length = int(dimensions[1])
            
            #make the maze_layout
            maze_layout = []
            for x in range(width):
                maze_layout.append([])
            
            for y in range(length):
                line = lines[y+1].split(" ")
                for x in range(width):
                    if line[x] != "":
                        state = int(line[x])
                        maze_layout[x].append(state)
            print("Ready reading maze file " + file_path)
            return Maze(maze_layout, width, length)
        except FileNotFoundError:
            print("Error reading maze file " + file_path)
            traceback.print_exc()
            sys.exit()